feat: add date picker to class modal

wais-hundekar opened the pull request:
## Notion ticket link
<!-- Please replace with your ticket's URL -->
[Update Create Class modal to match updated designs](https://www.notion.so/uwblueprintexecs/47bf4727a2444bccad4a91952898b8f3?v=9658436bf9be409e90581631a7155f16&p=1c95e8134c4e4c63b8644b3d124ffc07&pm=s)


<!-- Give a quick summary of the implementation details, provide design justifications if necessary -->
## Implementation description
* Updated backend to support `startDate` over `schoolYear`
* Implemented date picker from [chakra-dayzed-datepicker
](https://github.com/aboveyunhai/chakra-dayzed-datepicker)
* Resolved all resulting TypeScript errors and updated code to match designs
<img width="330" alt="Screenshot 2023-05-23 at 10 07 21 PM" src="https://github.com/uwblueprint/jump-math/assets/87676999/5ee801e8-47f2-4103-8de9-95d70b4a3d90">

![Screenshot 2023-05-21 at 10 21 17 AM](https://github.com/uwblueprint/jump-math/assets/87676999/3cac6f04-90fb-4a7e-a399-82b9556dcf70)

<!-- What should the reviewer do to verify your changes? Describe expected results and include screenshots when appropriate -->
## Steps to test
1. Test that mutation is successful
2. Ensure component matches design as closely as possible


<!-- Draw attention to the substantial parts of your PR or anything you'd like a second opinion on -->
## What should reviewers focus on?
* Is there any extra code I should/can remove?
* Any File Name changes?


## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

wais-hundekar committed:
schoolYear to startDate

wais-hundekar committed:
Merge branch 'staging' of https://github.com/uwblueprint/jump-math into wais/modal-date-picker

wais-hundekar committed:
date picker, modified tests and working mutation

wais-hundekar committed:
calender and styles

wais-hundekar committed:
remove unused code

wais-hundekar committed:
final design changes

wais-hundekar committed:
overridding css

wais-hundekar committed:
removing react-focus-lock

wais-hundekar commented:
Just carrying over @jfdoming comment from the closed PR:
"Aside from my current comments, I'm also a little wary of the hard-coded month/weekday names. Dates are tricky and i18n will be difficult with this system (if we ever wanted to do that)! If it's too much work to migrate to a browser API for this I'm fine leaving that part though."

I would be interested in discussing the time tradeoff to implement this, and if the new PR changes would affect this in anyway.

wais-hundekar committed:
Merge branch 'staging' of https://github.com/uwblueprint/jump-math into wais/modal-date-picker2

carissa-tang commented on frontend/src/components/user-management/student/AddClassroomModal.tsx:
```diff
@@ -151,13 +152,66 @@ const AddClassroomModal = ({
+                        "October",
+                        "November",
+                        "December",
+                      ], // length of 12
```
can we remove these comments?

carissa-tang commented on frontend/src/components/user-management/student/AddClassroomModal.tsx:
```diff
@@ -151,13 +152,66 @@ const AddClassroomModal = ({
-                    value={watch("schoolYear")}
+                  <FormLabel color="blue.300">Start Date</FormLabel>
+                  <SingleDatepicker
+                    configs={{
```
can we move this to a new datepicker component in `common`?

carissa-tang commented on frontend/src/components/user-management/student/AddClassroomModal.tsx:
```diff
@@ -151,13 +152,66 @@ const AddClassroomModal = ({
+                        "November",
+                        "December",
+                      ], // length of 12
+                      firstDayOfWeek: 1, // default is 0, the dayNames[0], which is Sunday if you don't specify your own dayNames,
```
wait this is design question but why does our calendar start on monday ?? 0_o

carissa-tang commented on frontend/src/components/user-management/student/AddClassroomModal.tsx:
```diff
@@ -151,13 +152,66 @@ const AddClassroomModal = ({
-                    type="number"
-                    value={watch("schoolYear")}
+                  <FormLabel color="blue.300">Start Date</FormLabel>
+                  <SingleDatepicker
```
can you add "Please choose a date" as a placeholder text?
<img width="239" alt="image" src="https://github.com/uwblueprint/jump-math/assets/69697180/f24464d9-ab3b-4aba-b706-3b7de4d4e6b0">

wais-hundekar committed:
day start on sunday

joyce-shi committed:
feat: student assessment experience (#282)

* add new route & page

* render some initial ui components

* add question number navigation

* cleanup

* beginning answers array logic

* a janky working implementation

* fix question numbers

* some cleanup

* fix answer index

* add multiple choice to question experience

* refactor function

* add support for multi select

* bug fixes and cleanup

* remove unused imports

* bug fixes

* refactor to use hook

* rename folder and utils

* oop

* pr comments: round 1

* pr comments: move question status to student utils

* pr comments: store isCompleted in Answers object

* pr comments: remove extraneous index fields

* pr comments: clean up answer element index function

* pr comments: rename assessment experience component

* pr comments: rename updatedAnswer function

* pr comments: remove unary operator usage

* fix question numbers

* add comments

* disable input field scrolling

jfdoming committed:
fix: update prettier to latest version (#345)

* Update prettier

This is to match the updated typescript from before.

* Use the new prettier version to auto-format the project files

* Auto-format new changes

wais-hundekar committed:
backend: Compute Test Completion Rate (#327)

* compute completion rate

* nit changes

jfdoming committed:
feat: corrected-short-answer component (#347)

* Add corrected-answer helper components

* Add corrected-short-answer component

* Add demo of short-answer component to component library

jfdoming committed:
feat: add student list component (#343)

* Add basic student list component

* Change sort icon to reflect designs (this is a global change)

* Fix background of search bar to not fade out

* Make list elements visually clickable

* Include `isViewed` attribute in student DTO (exact format tbd?)

* Allow selecting the current student

* Make student list searchable and sortable

* Add more test data

* Fix styling issues for overflowing list

* Use correct icon name

* Auto-format

jfdoming committed:
fix: remove unused dependency `rjsf` (#350)

Remove unused rjsf packages

jfdoming committed:
fix: upgrade `multer` to the latest stable version (#349)

Upgrade `multer` to the latest stable version

This should resolve a vulnerability Dependabot is complaining about.

jfdoming committed:
feat: multi-option components (#348)

* Add generic multi-option elements component

* Add multi-select component

* Add multiple-choice component

* Add demo of multi-option components to component library

jfdoming committed:
fix: organize data-viz package (#354)

Organize session data-viz package

joyce-shi committed:
feat: student test submission (#324)

* add new route & page

* render some initial ui components

* add question number navigation

* cleanup

* beginning answers array logic

* a janky working implementation

* fix question numbers

* some cleanup

* fix answer index

* add multiple choice to question experience

* refactor function

* add support for multi select

* bug fixes and cleanup

* remove unused imports

* bug fixes

* refactor to use hook

* rename folder and utils

* oop

* pr comments: round 1

* pr comments: move question status to student utils

* pr comments: store isCompleted in Answers object

* pr comments: remove extraneous index fields

* pr comments: clean up answer element index function

* add resolver for create test session result

* pr comments: rename assessment experience component

* add gql mutation

* add answer mapping util function

* update test session type file

* pr comments: rename updatedAnswer function

* pr comments: remove unary operator usage

* fix question numbers

* add comments

* disable input field scrolling

* use updated linter

* linter

* fix casting issues

* fix 0 index problem & add multi-select array sorting

* add handling on submission

* cleanup

* split student context

* update confirmation modal to reflect unanswered questions count

* fix handler function

* addressing pr comments

* pr comments - rename variable

joyce-shi committed:
feat: add loading state to student login (#351)

* add loader

* add default theming to spinner

wais-hundekar committed:
comments and placeholder

wais-hundekar committed:
refactoring datepicker

dependabot[bot] committed:
[npm backend] (deps-dev): Bump @types/cors from 2.8.10 to 2.8.13 in /backend (#357)

[npm backend] (deps-dev): Bump @types/cors in /backend

Bumps [@types/cors](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/cors) from 2.8.10 to 2.8.13.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/cors)

---
updated-dependencies:
- dependency-name: "@types/cors"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
[npm backend] (deps-dev): Bump @types/lodash from 4.14.194 to 4.14.195 in /backend (#358)

[npm backend] (deps-dev): Bump @types/lodash in /backend

Bumps [@types/lodash](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/lodash) from 4.14.194 to 4.14.195.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/lodash)

---
updated-dependencies:
- dependency-name: "@types/lodash"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
[npm backend] (deps-dev): Bump @types/validator from 13.7.15 to 13.7.17 in /backend (#359)

[npm backend] (deps-dev): Bump @types/validator in /backend

Bumps [@types/validator](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/validator) from 13.7.15 to 13.7.17.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/validator)

---
updated-dependencies:
- dependency-name: "@types/validator"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
[npm backend] (deps): Bump cookie-parser and @types/cookie-parser in /backend (#361)

[npm backend] (deps): Bump cookie-parser and @types/cookie-parser

Bumps [cookie-parser](https://github.com/expressjs/cookie-parser) and [@types/cookie-parser](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/cookie-parser). These dependencies needed to be updated together.

Updates `cookie-parser` from 1.4.5 to 1.4.6
- [Release notes](https://github.com/expressjs/cookie-parser/releases)
- [Changelog](https://github.com/expressjs/cookie-parser/blob/master/HISTORY.md)
- [Commits](https://github.com/expressjs/cookie-parser/compare/1.4.5...1.4.6)

Updates `@types/cookie-parser` from 1.4.2 to 1.4.3
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/cookie-parser)

---
updated-dependencies:
- dependency-name: cookie-parser
  dependency-type: direct:production
  update-type: version-update:semver-patch
- dependency-name: "@types/cookie-parser"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
[npm backend] (deps): Bump sequelize-typescript from 2.1.0 to 2.1.5 in /backend (#363)

[npm backend] (deps): Bump sequelize-typescript in /backend

Bumps [sequelize-typescript](https://github.com/RobinBuschmann/sequelize-typescript) from 2.1.0 to 2.1.5.
- [Release notes](https://github.com/RobinBuschmann/sequelize-typescript/releases)
- [Changelog](https://github.com/sequelize/sequelize-typescript/blob/master/CHANGELOG.md)
- [Commits](https://github.com/RobinBuschmann/sequelize-typescript/compare/v2.1.0...v2.1.5)

---
updated-dependencies:
- dependency-name: sequelize-typescript
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
[npm frontend] (deps): Bump web-vitals from 1.1.0 to 3.3.1 in /frontend (#362)

Bumps [web-vitals](https://github.com/GoogleChrome/web-vitals) from 1.1.0 to 3.3.1.
- [Changelog](https://github.com/GoogleChrome/web-vitals/blob/main/CHANGELOG.md)
- [Commits](https://github.com/GoogleChrome/web-vitals/compare/v1.1.0...v3.3.1)

---
updated-dependencies:
- dependency-name: web-vitals
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

joyce-shi committed:
fix: reload prompt (#356)

* fix reload

* fix reload hook

dependabot[bot] committed:
[npm frontend] (deps-dev): Bump eslint-plugin-jsx-a11y from 6.4.1 to 6.7.1 in /frontend (#365)

[npm frontend] (deps-dev): Bump eslint-plugin-jsx-a11y in /frontend

Bumps [eslint-plugin-jsx-a11y](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y) from 6.4.1 to 6.7.1.
- [Release notes](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/releases)
- [Changelog](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/CHANGELOG.md)
- [Commits](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/compare/v6.4.1...v6.7.1)

---
updated-dependencies:
- dependency-name: eslint-plugin-jsx-a11y
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Julian Dominguez-Schatz <julian.dominguezschatz@gmail.com>
Co-authored-by: Joyce Shi <joyce.shi@uwaterloo.ca>

dependabot[bot] committed:
[npm frontend] (deps-dev): Bump @typescript-eslint/eslint-plugin from 5.59.6 to 5.59.8 in /frontend (#367)

[npm frontend] (deps-dev): Bump @typescript-eslint/eslint-plugin

Bumps [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/eslint-plugin) from 5.59.6 to 5.59.8.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v5.59.8/packages/eslint-plugin)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/eslint-plugin"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
[npm frontend] (deps-dev): Bump @typescript-eslint/parser from 5.59.6 to 5.59.8 in /frontend (#368)

[npm frontend] (deps-dev): Bump @typescript-eslint/parser in /frontend

Bumps [@typescript-eslint/parser](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/parser) from 5.59.6 to 5.59.8.
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/parser/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v5.59.8/packages/parser)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/parser"
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

carissa-tang committed:
feat: Return StudentDTOs in testSession Resolver (#346)

d282liu committed:
feat: fetch data for classroom page (#353)

* init

* removed unnecessary import

* addressed pr comments

* moved ErrorStateWithButton to common

* moved ErrorStateWithButton out of common

---------

Co-authored-by: David Liu <davidliu@Davids-MacBook-Air.local>
Co-authored-by: David Liu <davidliu@v1040-wn-rt-a-71-56.campus-dynamic.uwaterloo.ca>

jfdoming committed:
feat: View Student Result page (#355)

* Add common `RouterTabs` component

* Add basic structure for new page

* Add basic student page structure

* Add statistics card section

* Add student answers section template

* Resize student list container

* Fix various padding issues, plus add missing title

* Linter

* Linter, pt. 2

* Fix focus ring clipping

* Fix container width

* PR feedback: remove unnecessary padding

* PR feedback: use correct divider colour

* PR feedback: use correct fallback

* PR feedback: reuse back button component

* Implement back-navigation

* PR feedback: extract utility file

* PR feedback: improve file structure

* lint: no-explicit-any

* PR feedback: move statistics utils

joyce-shi committed:
feat: distribute assessment flow wrapper (#352)

* initial commit

* build distribute assessment flow wrapper

* add breadcrumb header

* add links to breadcrumb

* fix active link issue

* cleanup

* pr comment - top-level config

joyce-shi committed:
chore: organize folders (#371)

* fix empty classrooms state

* move message components

* more folder reorganization

* more folder reorg

* organize student folder

joyce-shi committed:
chore: reorganize folders again (#373)

reorganize folders again

jfdoming committed:
fix: remove unused bootstrap dependency (#374)

jfdoming committed:
fix: complete react version upgrade (#376)

* Update types to the matching version

* Fix React 17 -> 18 discrepancies in code

* Upgrade react-router v5 to a version compatible with React 18

joyce-shi commented on backend/models/class.model.ts:
```diff
@@ -12,7 +12,7 @@ export interface Class extends Document {
   /** the name of the class */
   className: string;
   /** the school year of the class */
```
could you update the comment? thanks!

joyce-shi commented on frontend/src/components/user-management/student/AddClassroomModal.tsx:
```diff
@@ -151,13 +155,10 @@ const AddClassroomModal = ({
+                  <FormLabel color="blue.300">Start Date</FormLabel>
+                  <DatePicker
+                    onChange={handleDateChange}
+                    value={watch("startDate")}
```
do we check somewhere that the `startDate` is non-empty and on or after the current date?

joyce-shi reviewed (commented):
could you merge master? i think there might be some conflicts. then i will approve!

carissa-tang committed:
fix: miscellaneous image bugs (#370)

jfdoming committed:
fix: crashing heroku deployment (#378)

* fix: Copy `.cjs` files to `build/` directory after `yarn build`

* fix: Remove the need for `.cjs` extensions for library files

* fix: Add eslint rule to catch manual imports of `.cjs` files

* fix: Ignore build directory when using nodemon

* fix: Lint ALL project files (not just `.ts`/`.js` files)

jfdoming committed:
fix: miscellaneous bugfixes (#379)

* Test session codes should not have a space when copied

* Tabs should change immediately when clicked, not upon release

* `Switch`es should only have routes as children

* `key` props should be meaningful and unique

* `onCompleted` should use the data it is provided

* Fix modals not closing properly

Due to the React 18 version upgrade, framer-motion needed to be
upgraded, or else all our models would not close properly.

More info: https://github.com/chakra-ui/chakra-ui/issues/5855#issuecomment-1101521198

* Fix invalid DOM-nesting

https://www.notion.so/uwblueprintexecs/Fix-invalid-nested-button-markup-in-TestSessionListItem-component-ed52aeb4fa4147a79e47baaadfa80e58

* Make short-answer input fully controlled

* Standardize tab theming

wais-hundekar committed:
Revert "merge conflicts"

This reverts commit c905a75ffcc9bc346acb0e55c7d2ce672abbd0f3, reversing
changes made to f01a0b9e28ac2be562d3cf00f62f5db1f5c26559.

wais-hundekar committed:
Merge branch 'staging' of https://github.com/uwblueprint/jump-math into wais/modal-date-picker2
